src/librustc_mir/build/expr/as_rvalue.rs

```diff
@@ -433,55 +438,69 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
 
         let arg_place = unpack!(block = this.as_place(block, arg));
 
-        let mutability = match arg_place {
-            Place::Local(local) => this.local_decls[local].mutability,
-            Place::Projection(box Projection {
-                base: Place::Local(local),
-                elem: ProjectionElem::Deref,
-            }) => {
-                debug_assert!(
-                    if let Some(ClearCrossCrate::Set(BindingForm::RefForGuard))
-                        = this.local_decls[local].is_user_variable {
-                        true
-                    } else {
-                        false
+        let mut mutability = if let PlaceBase::Local(local) = arg_place.base {
+            this.local_decls[local].mutability
+        } else {
+            bug!("Unexpected capture place");
+        };
+        if !arg_place.has_no_projection() {
+            let mut proj_iter = arg_place.elems.iter().cloned().rev();
+            if let Some(projection) = proj_iter.next() {
+                match projection {
+                    ProjectionElem::Deref => {
+                        if let PlaceBase::Local(local) = arg_place.base {
+                            debug_assert!(
+                                if let Some(ClearCrossCrate::Set(BindingForm::RefForGuard))
+                                = this.local_decls[local].is_user_variable {
+                                    true
+                                } else {
+                                    false
+                                },
+                                "Unexpected capture place",
```

Comment:
I'm currently getting a failure here, and I think it's because this code is not really equivalent to the old code. I think this is where a `split_projection` sort of thing would be useful. In any case, the problem here I believe:

If you have something like `*x.f` -- this has a tree structure like:

- deref
  - field (`f`)
    - base: `x`

but in the code here we will pull off the last projection (`*`) and then look at `arg_place.base`, which is `x`, ignoring the field `f` that comes in between.